src/SparkConduit.sol

```diff
@@ -0,0 +1,314 @@
+
+        address source = RegistryLike(registry).buffers(ilk);
+
+        require(
+            IERC20(asset).transferFrom(source, address(this), amount),
+            "SparkConduit/transfer-failed"
+        );
+
+        pool.supply(asset, amount, address(this), 0);
+
+        // Convert asset amount to shares
+        uint256 shares = amount.rayDiv(pool.getReserveData(asset).liquidityIndex);
+
+        assets[asset].positions[ilk].shares += shares;
+        assets[asset].totalShares           += shares;
+
+        emit Deposit(ilk, asset, source, amount);
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function withdraw(bytes32 ilk, address asset, uint256 maxAmount)
+        external ilkAuth(ilk) returns (uint256 amount)
+    {
+        uint256 liquidityAvailable
+            = IERC20(asset).balanceOf(pool.getReserveData(asset).aTokenAddress);
+
+        amount = liquidityAvailable < maxAmount ? liquidityAvailable : maxAmount;
+
+        // Normally you should update local state first for re-entrancy,
+        // but we need an update-to-date liquidity index for that
+        amount = pool.withdraw(asset, amount, address(this));
+
+        // Convert asset amount to shares
+        uint256 liquidityIndex = pool.getReserveData(asset).liquidityIndex;
+        uint256 shares         = amount.rayDiv(liquidityIndex);
+
+        uint256 pshares     = assets[asset].positions[ilk].shares;
+        uint256 withdrawals = assets[asset].positions[ilk].pendingWithdrawals;
+
+        if (pshares < shares) {
+            uint256 toReturnShares;
+
+            unchecked {
+                toReturnShares = shares - pshares;
+            }
+
+            uint256 toReturn = toReturnShares.rayMul(liquidityIndex);
+            amount -= toReturn;
+            shares = pshares;
+
+            // Return the excess to the pool as it's other user's deposits
+            // Round against the user that is withdrawing
+            pool.supply(asset, toReturn, address(this), 0);
+        }
+
+        assets[asset].positions[ilk].shares -= shares;
+        assets[asset].totalShares           -= shares;
+
+        if (withdrawals > 0) {
+            if (shares <= withdrawals) {
+                assets[asset].positions[ilk].pendingWithdrawals -= shares;
+                assets[asset].totalPendingWithdrawals           -= shares;
+            } else {
+                assets[asset].positions[ilk].pendingWithdrawals = 0;
+                assets[asset].totalPendingWithdrawals           -= withdrawals;
+            }
+        }
+
+        address destination = RegistryLike(registry).buffers(ilk);
+
+        IERC20(asset).transfer(destination, amount);
+
+        emit Withdraw(ilk, asset, destination, amount);
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function maxDeposit(bytes32, address asset) public view returns (uint256 maxDeposit_) {
+        // Note: Purposefully ignoring any potental supply cap limits on Spark
```

Comment:
Why are we ignoring here? It will revert on supply if we hit the cap

Reply:
This was similar to a decision made in the D3M. Aave V3 supports supply caps so yes this would cause an error if it's not respected, but we just make the assumption there is no supply cap here as it makes the calculation way simpler. There's no plan to have a supply cap for DAI especially as it can't be used as collateral it doesn't really matter.

Reply:
`potential`, let's state the assumption above the function with a quick justification

Reply:
expanded on it